Add temporary impl_UNBOXED syntax sugar for unboxed-only defs.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35714 Add temporary impl_UNBOXED syntax sugar for unboxed-only defs.**
* #35706 Add DispatchKey impl overload; remove use of torch::dispatch
* #35476 Add unboxedCallRedispatch

There are a lot of unboxed only defs.  We're committed to removing
them at the end of the half but as I am about to do a lot of porting
to the new API, let's get them into a form where they're easy to
remove.  This is a new overload impl_UNBOXED that will pass
the function pointer straight to CppFunction::makeUnboxedOnly

I don't attempt to make the _UNBOXED API complete; in particular,
catchall declarations don't get this sugar (as there are very few
of them).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20775782](https://our.internmc.facebook.com/intern/diff/D20775782)

ezyang committed:
Add temporary impl_UNBOXED syntax sugar for unboxed-only defs.

There are a lot of unboxed only defs.  We're committed to removing
them at the end of the half but as I am about to do a lot of porting
to the new API, let's get them into a form where they're easy to
remove.  This is a new overload impl_UNBOXED that will pass
the function pointer straight to CppFunction::makeUnboxedOnly

I don't attempt to make the _UNBOXED API complete; in particular,
catchall declarations don't get this sugar (as there are very few
of them).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/op_registration/op_registration.h:
```diff
@@ -879,6 +879,23 @@ class CAFFE2_API Module final {
+  // CppFunction::makeUnboxedOnly factory directly (the API is compositional;
+  // it's just that we have a LOT of impl calls that are unboxed, so there is
+  // just a little syntax sugar for this case.)
+  template <typename Dispatch, typename Func>
```
Can we add a TODO here saying we intend to remove these?

ezyang replied:
sure

ezyang committed:
Update on "Add temporary impl_UNBOXED syntax sugar for unboxed-only defs."


There are a lot of unboxed only defs.  We're committed to removing
them at the end of the half but as I am about to do a lot of porting
to the new API, let's get them into a form where they're easy to
remove.  This is a new overload impl_UNBOXED that will pass
the function pointer straight to CppFunction::makeUnboxedOnly

I don't attempt to make the _UNBOXED API complete; in particular,
catchall declarations don't get this sugar (as there are very few
of them).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add temporary impl_UNBOXED syntax sugar for unboxed-only defs."


There are a lot of unboxed only defs.  We're committed to removing
them at the end of the half but as I am about to do a lot of porting
to the new API, let's get them into a form where they're easy to
remove.  This is a new overload impl_UNBOXED that will pass
the function pointer straight to CppFunction::makeUnboxedOnly

I don't attempt to make the _UNBOXED API complete; in particular,
catchall declarations don't get this sugar (as there are very few
of them).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20775782](https://our.internmc.facebook.com/intern/diff/D20775782)

[ghstack-poisoned]
